sympy/functions/elementary/integers.py

```diff
@@ -31,7 +31,12 @@ def eval(cls, arg):
             return v
 
         if arg.is_integer or arg.is_finite is False:
-            return arg
+            if arg.is_real is False :
```

Comment:
Can this case actually be trigged?

`(2*I).is_integer` returns  `False`, so what remains are infinite numbers. Probably something like:
```
if arg.is_integer:
    if arg.is_Number:
         return arg
    else:
        return arg.cancel()
if arg.is_finite is False:
   return arg
```
is more efficient. (I think `is_Number` is faster than `is_number`.)

Or do you have any examples which hit all three branches?

Reply:
I had involved three conditions since there were a couple of fails related to imaginary numbers and one simplify fail in 'test_diophantine.py` which returned the simplified answer which was unexpected .
```
assert parametrize_ternary_quadratic(eq) == \
        (8*p**2 - 3*p*q, -8*p*q + 8*q**2, 5*p*q)
```
I have understood the changes you suggest and also ran tests .I will look into the suggested expression which fails according to you currently and then make a commit along with tests .

Reply:
If you get a *better* results after a change, feel free to update the test (it will be reviewed as well).

Reply:
I thought a bit more about it and I assume that the reason for returning arg here is that we knows that we have the final result, so an early exit.

I would change the original condition to 
```
if arg.is_Atom and arg.is_integer or arg.is_finite is False:
```

(which should be the same as `if arg.is_Atom and arg.is_integer or arg.is_infinite:`)

So we exit here it the value is infinite or if it is a single number/symbol and integer.
This may to some extent solve the problem (as we do not exit because it `is_integer`, but not a single value).

Here one can consider performing a cancel and then do another similar test, in case something evaluates to the something where we can do an early exit.

Reply:
Yeah I will look into this .Thanks for the suggestions @oscargus 
I will make the changes soon !